src/codegen/spirv.zig

```diff
@@ -1750,24 +1761,32 @@ const NavGen = struct {
                 var member_names = std.ArrayList([]const u8).init(self.gpa);
                 defer member_names.deinit();
 
+                var index: u32 = 0;
+                var offset: u32 = 0;
                 var it = struct_type.iterateRuntimeOrder(ip);
-                while (it.next()) |field_index| {
+                const result_id = self.spv.allocId();
+                while (it.next()) |field_index| : (index += 1) {
                     const field_ty = Type.fromInterned(struct_type.field_types.get(ip)[field_index]);
                     if (!field_ty.hasRuntimeBitsIgnoreComptime(zcu)) {
                         // This is a zero-bit field - we only needed it for the alignment.
                         continue;
                     }
 
+                    try self.spv.decorateMember(result_id, index, .{ .Offset = .{ .byte_offset = offset } });
                     const field_name = struct_type.fieldName(ip, field_index).unwrap() orelse
                         try ip.getOrPutStringFmt(zcu.gpa, pt.tid, "{d}", .{field_index}, .no_embedded_nulls);
                     try member_types.append(try self.resolveType(field_ty, .indirect));
                     try member_names.append(field_name.toSlice(ip));
+
+                    offset += @intCast(field_ty.abiSize(zcu));
```

Comment:
This computation can be done more rigorously by `struct_type.structFieldOffset`. The LLVM backend also does a bunch of weird stuff, but this is probably good enough for now.

Reply:
Oh also, the `Offset` decoration is only allowed for Vulkan and not for OpenCL. Can you add a check?